Front display course

LucasDumestre opened the pull request.

LucasDumestre committed:
Delete front end react app

LucasDumestre committed:
Add new file in the backend to manage the front

LucasDumestre committed:
Modify calendar json data

LucasDumestre committed:
Add comments in front code

LucasDumestre committed:
New template for the landing page

bastantoine commented on API/templates/template.html:
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>Mappy Telecom</title>
```
On avait pas dit _Mappy'MT_ ? 😇

bastantoine commented on API/front.py:
```diff
@@ -0,0 +1,27 @@
+@front.route('/')
+def template_test():
+    # We get the user information from our mocked API
+    user = requests.get('https://hackathoooon.osc-fr1.scalingo.io/mock/pass/calendar?user=1')
```
En général on barcode pas l'entièreté de l'url dans le code, sauf si on est absolument certain qu'on veut utiliser la même quelque soit ton env. Ce que tu peux faire c'est mettre l'option dans un fichier `.env` et l'utiliser ensuite. Pour ça :

1. Tu crée un fichier `.env` (comme ça, juste avec l'extension) dans le dossier courant. 
  - Ce fichier là doit pas être track par git puisqu'il permet d'avoir des configs différentes selon les environnements utilisés. Comme ça quand tu lance en local tu peux taper une API en local et quand c'est lance en prod, ça tape l'API de prod.
  - Pour que quelqu'un qui configure le projet pour la première fois sache ce qu'il doit configurer, tu peux faire une copie de ton fichier `.env` en `.env.sample`, qui lui sera track par git
2. Dans ce fichier tu ajoutes sur une nouvelle ligne où tu défini une variable, genre `API_URL` (par convention, les variables d'env on les mets en majuscule) à laquelle tu assigne un string avec l'url du root de ton API.
3. Tu installes le module `dotenv`, et commit l'ajout dans le `requirements.txt`:
```bash
$ pip install python-dotenv
$ pip freeze > requirements.txt
$ git add requirements.txt && git commit -m 'pip install python-dotenv'
```
4. Au début de ton fichier, tu rajoutes :
```python
from dotenv import load_dotenv
load_dotenv()
```
5. Ensuite, au moment d'effectuer la requête :
```python
user = requests.get(os.path.join(os.environ.get('API_URL', ''), 'mock/pass/calendar?user=1'))
```
(En oubliant pas le `import os` au début du fichier évidemment).

bastantoine commented on API/resources/pass.json:
```diff
@@ -3,54 +3,54 @@
     "firstname": "Lucas",
     "calendar": {
-      "01-19-2021": [
+      "Tuesday": [
```
La semaine commence le mardi ? 😅 Je veux bien qu'on ait plus eu de cours le lundi depuis longtemps, mais jadis on en a eu.

bastantoine commented on API/front.py:
```diff
@@ -0,0 +1,27 @@
+        except:
+            # We pass to the next day
+            today += datetime.timedelta(days=1)
+            break
```
Pourquoi un `try/except` ici ? S'il a une réelle utilité, il vaut mieux voir précisément quelle exception tu essaie de catch et l'indiquer dans le `except`, autrement avec ça tu catch toutes les exceptions sans les traiter correctement derrière, ça peut engendrer des problèmes potentiels.

Par ailleurs, pourquoi tu break dans ton except ? Actuellement ton `while` sert à rien puisque dans tous les cas tu break à la première itération.

LucasDumestre replied:
Enfaite au début je voulais rechercher le cours suivant dans le calendrier. Dans la journée ou le prochain dans les jours suivants.

Mais je ne voyais pas trop comment parcourir les horaire en python autrement qu'avec while, puis pour le prototype, je me suis dis qu'on prenais juste le premier cours de la journée pour la prochaine journée qui a un cours.

bastantoine replied:
Utilise ça :
```python
def get_current_or_next_course():
    user_calendar = json.loads(...)['calendar']

    # We search for the next course
    now = datetime.now()
    today = now.strftime('%A')

    # First try to get the courses of today
    today_courses = user_calendar.get(today, [])

    if today_courses:
        # We have one or more classes this day
        next_or_current_courses = list(filter(
            lambda course: now < datetime.strptime(course['end'], '%H:%M'),
            today_courses
        ))
        if next_or_current_courses:
            # Return the next or current course left in the day
            return next_or_current_courses[0]

    next_day = now
    while next_day < next_day + timedelta(days=7):
        next_day += timedelta(days=1)
        this_day_courses = user_calendar.get(next_day.strftime('%A'), [])
        if this_day_courses:
            return this_day_courses[0]

    return {}
```

bastantoine commented on API/resources/pass.json:
```diff
@@ -3,54 +3,54 @@
+          "location": "B01-014A"
         }
       ],
-      "01-20-2021": [
```
Pourquoi avoir viré les dates ?

LucasDumestre replied:
J’ai viré les dates des jours car comme ça si on laisse le prototype up chaque semaine on bouclera sur les cours défini

bastantoine commented on API/front.py:
```diff
@@ -0,0 +1,27 @@
+
+    # We search the next_course
+    today = datetime.today()
+    today_course = ""
```
Si tu trouve pas de cours pour aujourd'hui, `today_course` sera toujours un string vide quand tu vas faire le rendu de ton template, et ça va planter.
```suggestion
    today_course = {}
```

bastantoine commented on API/templates/template.html:
```diff
@@ -0,0 +1,17 @@
+  </head>
+  <body>
+    <div class="alert alert-success" role="alert">
+        Votre prochain cours de {{ today[0].name }}<br>aura lieu en {{ today[0].location }}<br>Il commencera à {{ today[0].start }}
```
Tu traites pas de cas où on trouve pas de cours pour le jour courant, alors que tu gères ce cas (où du moins essaie 😉) dans ta vue.

bastantoine commented on API/templates/template.html:
```diff
@@ -0,0 +1,17 @@
+  <head>
+    <title>Mappy Telecom</title>
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <script src="http://code.jquery.com/jquery-1.10.2.min.js"></script>
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
+    <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
```
Tu load deux fois jQuery avec des versions différentes, et très éloignées l'une de l'autre.

bastantoine commented on API/main.py:
```diff
@@ -5,15 +5,14 @@
 
-@app.route('/')
-def main():
-    return 'Hello there!'
```
RIP Obi 😢

bastantoine commented on API/main.py:
```diff
@@ -30,6 +29,8 @@ def pass_controller():
     calendar = data.get(user, {})
     return calendar
 
+import front
```
Pourquoi l'importer ici alors que tu le fais déjà au début ? Par ailleurs, [par convention](https://www.python.org/dev/peps/pep-0008/#imports), en Python tous les imports se font en début de fichier, sauf rares exceptions.

LucasDumestre committed:
Some correction following the review

- Mappy'MT
- env for API url
- replace var today_course init by {} instead of ""
- Remove old jquery
- Remove duplicate import in main.py

LucasDumestre committed:
Modify front to get and display the next scheduled class

bastantoine reviewed (changes requested):
Bon pour moi, je vais mettre à jour les workflows et puis je mergerais

bastantoine committed:
Removed outdated docker images

bastantoine committed:
Removed useless GitHub Action workflow files
